### PR TITLE
Classic nuts fixes

### PR DESCRIPTION
**Pull request by gordoncaleb**
A few fixes to our current NUTS implementation. It fixes:

1. Leapfrog was not leaping properly
1. The tree size and sum acceptance rate wasn't being tracked correctly
1. NUTS now checks to a network in a zero probability state before running inference

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Leapfrog.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public Leapfrog step(final List<? extends Variable<DoubleTensor, ?>> latentVaria
         final double halfTimeStep = epsilon / 2.0;
 
         Map<VariableReference, DoubleTensor> nextMomentum = stepMomentum(halfTimeStep, momentum, gradient);
-        Map<VariableReference, DoubleTensor> nextPosition = stepPosition(latentVariables, halfTimeStep, nextMomentum, position);
+        Map<VariableReference, DoubleTensor> nextPosition = stepPosition(latentVariables, epsilon, nextMomentum, position);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Leapfrog.java]: This is huge

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Leapfrog.java [hunk 2]
```diff
@@ -58,16 +58,14 @@ public Leapfrog step(final List<? extends Variable<DoubleTensor, ?>> latentVaria
         return new Leapfrog(nextPosition, nextMomentum, nextPositionGradient);
     }
 
-    private Map<VariableReference, DoubleTensor> stepPosition(List<? extends Variable<DoubleTensor, ?>> latentVariables, double halfTimeStep, Map<VariableReference, DoubleTensor> nextMomentum, Map<? extends VariableReference, DoubleTensor> position) {
+    private Map<VariableReference, DoubleTensor> stepPosition(List<? extends Variable<DoubleTensor, ?>> latentVariables, double dt, Map<VariableReference, DoubleTensor> nextMomentum, Map<? extends VariableReference, DoubleTensor> position) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Leapfrog.java]: No real changes other than a rename

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTS.java [hunk 3]
```diff
@@ -98,13 +100,20 @@ private NUTSSampler setupSampler(final ProbabilisticModelWithGradient model,
...
-        Map<VariableReference, DoubleTensor> startingSample = SamplingAlgorithm.takeSample(latentVariables);
-        Map<VariableReference, DoubleTensor> position = startingSample.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> (DoubleTensor) e.getValue()));
-        Map<VariableReference, DoubleTensor> momentum = new HashMap<>();
-        Map<? extends VariableReference, DoubleTensor> gradient = model.logProbGradients();
+        Map<VariableReference, DoubleTensor> position = latentVariables.stream()
+            .collect(toMap(Variable::getReference, Variable::getValue));
 
         double initialLogOfMasterP = model.logProb(position);
 
+        Preconditions.checkArgument(
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTS.java]: Check for zero prob first

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSSampler.java [hunk 4]
```diff
@@ -127,8 +127,8 @@ public void step() {
             }
 
             tree.incrementLeapfrogCount(otherHalfTree.getAcceptedLeapfrogCount());
-            tree.setDeltaLikelihoodOfLeapfrog(otherHalfTree.getDeltaLikelihoodOfLeapfrog());
-            tree.setTreeSize(otherHalfTree.getTreeSize());
+            tree.setDeltaLikelihoodOfLeapfrog(tree.getDeltaLikelihoodOfLeapfrog() + otherHalfTree.getDeltaLikelihoodOfLeapfrog());
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSSampler.java]: Sum the tree size and sum accept with the other half of tree

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Stepsize.java [hunk 5]
```diff
@@ -156,6 +162,6 @@ public double getStepsize() {
     @Override
     public void save(Statistics statistics) {
         statistics.store(NUTS.Metrics.STEPSIZE, stepsize);
-        statistics.store(NUTS.Metrics.MEAN_TREE_ACCEPT, averageAcceptanceProb);
+        statistics.store(NUTS.Metrics.MEAN_TREE_ACCEPT, averageTreeAcceptanceProb);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Stepsize.java]: Return the mean tree accept for stats not the H from the adapt equation

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/LeapfrogTest.java [hunk 6]
```diff
@@ -71,8 +71,8 @@ public void canLeapForward() {
         Leapfrog start = new Leapfrog(position, momentum, gradient);
         Leapfrog leap = start.step(vertices, mockedGradientCalculator, EPSILON);
 
-        Assert.assertEquals(1.0, leap.getPosition().get(aID).scalar(), 1e-6);
-        Assert.assertEquals(1.0, leap.getPosition().get(bID).scalar(), 1e-6);
+        Assert.assertEquals(2.0, leap.getPosition().get(aID).scalar(), 1e-6);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/LeapfrogTest.java]: Leaping correctly means the position should progress to 1*gradient not 0.5*gradient.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java [hunk 7]
```diff
@@ -76,20 +77,19 @@ public void canRecordStatisticsFromSamples() {
     public void samplesGaussian() {
         double mu = 0.0;
         double sigma = 1.0;
-        BayesianNetwork simpleGaussian = MCMCTestDistributions.createSimpleGaussian(mu, sigma, 3, random);
+        BayesianNetwork simpleGaussian = MCMCTestDistributions.createSimpleGaussian(mu, sigma, 3);
         ProbabilisticModelWithGradient model = new KeanuProbabilisticModelWithGradient(simpleGaussian);
 
+        int sampleCount = 1000;
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java]: Need fewer samples with the leapfrog correction

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java [hunk 8]
```diff
@@ -102,11 +102,9 @@ public void samplesContinuousPrior() {
         BayesianNetwork bayesNet = MCMCTestDistributions.createSumOfGaussianDistribution(20.0, 1.0, 46., 15.0);
         ProbabilisticModelWithGradient model = new KeanuProbabilisticModelWithGradient(bayesNet);
 
-        int sampleCount = 6000;
+        int sampleCount = 500;
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java]: Need way fewer samples

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java [hunk 9]
```diff
@@ -168,6 +165,22 @@ public void canDefaultToSettingsInBuilder() {
         assertFalse(posteriorSamples.get(A).asList().isEmpty());
     }
 
+    @Test(expected = IllegalArgumentException.class)
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java]: Add tests for zero prob network

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java [hunk 10]
```diff
@@ -196,12 +208,28 @@ public void checksSamplesAgainstMagicNumbers() {
 
         List<DoubleTensor> samples = posteriorSamples.get(vertex).asList();
 
-        Assert.assertEquals(3.0, samples.get(0).scalar(), 1e-9);
-        Assert.assertEquals(3.0, samples.get(1).scalar(), 1e-9);
-        Assert.assertEquals(0.9374092571432446, samples.get(2).scalar(), 1e-9);
-        Assert.assertEquals(0.05720950629236243, samples.get(3).scalar(), 1e-9);
-        Assert.assertEquals(0.33119352888492626, samples.get(4).scalar(), 1e-9);
-        Assert.assertEquals(0.9124861769925321, samples.get(19).scalar(), 1e-9);
+        double epsilon = 1e-9;
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/NUTSTest.java]: Add all steps in 20 samples
